beacon-chain/db/kv/state.go

```diff
@@ -120,9 +164,47 @@ func (s *Store) SaveStates(ctx context.Context, states []iface.ReadOnlyBeaconSta
 			if err := bucket.Put(rt[:], multipleEncs[i]); err != nil {
 				return err
 			}
+			if err := valIdxBkt.Put(rt[:], validatorKeys[i]); err != nil {
+				return err
+			}
+		}
+
+		// store the validator entries separately to save space.
+		valBkt := tx.Bucket(stateValidatorsBucket)
+		for hashStr, validatorEntry := range validatorsEntries {
+			key := hexutils.HexToBytes(hashStr)
+			// if the entry is not in the cache and not in the DB,
+			// then insert it in the DB and add to the cache.
+			if _, ok := s.validatorEntryCache.Get(key); !ok {
+				validatorEntryCacheMiss.Inc()
+				if valEntry := valBkt.Get(key); valEntry == nil {
+					valBytes, encodeErr := encode(ctx, validatorEntry)
+					if encodeErr != nil {
+						return encodeErr
+					}
+					if putErr := valBkt.Put(key, valBytes); putErr != nil {
+						return putErr
+					}
+					s.validatorEntryCache.Set(key, validatorEntry, int64(len(valBytes)))
+				}
+			} else {
```

Comment:
What about the possibility that the validator still exists in the cache but no longer exists in the DB? In this case it will not be saved. I think this scenario can happen because you do not remove the cache key inside `DeleteState`.

Reply:
Good point. When we delete a `state`, we effectively delete the `validator index` of that state. The `validator entries` are not deleted as such, since they will be used in some other state. Because of this very reason, they are not removed from cache. If the cache is full, they will be evicted based on the cost.

Reply:
It was decided in [discord](https://discordapp.com/channels/476244492043812875/550365651118850081/866975036489334804)  That we would delete the cache entry when we delete a state. This will have some effects on caching.  But it takes care f some corner case.